### PR TITLE
ADO: Continuous Assurance

### PR DESCRIPTION
**Pull request by Garima-msft**
## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+		$this.TimeStamp = (Get-Date -format "yyMMddHHmmss")
+		$this.StorageName = "adoscannersa"+$this.TimeStamp 
+		$this.FuncAppName = $this.FuncAppDefaultName + $this.TimeStamp 
+		$this.KeyVaultName = $this.KVDefaultName+$this.TimeStamp 
+		$this.AppInsightsName = $this.FuncAppName
+        $this.SetupComplete = $false
+        $this.ScanTriggerTimeUTC = [System.DateTime]::UtcNow.AddMinutes(20)
+		$this.ScanTriggerLocalTime = $(Get-Date).AddMinutes(20)
+		$this.ControlSettings = [ConfigurationManager]::LoadServerConfigFile("ControlSettings.json");
+		$this.ImageName = $this.ControlSettings.DockerImage.ImageName
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:64]: add checkmember

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+		$this.ProjectNames = $Proj
+		$this.ExtendedCommand = $ExtCmd
+		$this.TimeStamp = (Get-Date -format "yyMMddHHmmss")
+		$this.StorageName = "adoscannersa"+$this.TimeStamp 
+		$this.FuncAppName = $this.FuncAppDefaultName + $this.TimeStamp 
+		$this.KeyVaultName = $this.KVDefaultName+$this.TimeStamp 
+		$this.AppInsightsName = $this.FuncAppName
+        $this.SetupComplete = $false
+        $this.ScanTriggerTimeUTC = [System.DateTime]::UtcNow.AddMinutes(20)
+		$this.ScanTriggerLocalTime = $(Get-Date).AddMinutes(20)
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:62]: use this.timestamp

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 (hunk 1 above)
Comment (abhaydaga) [reply src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:64]: if not found - If you are using customized org policy, please ensure <varName> is defined in your ControlSettings.json

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+		[InvocationInfo] $invocationContext) : Base($subscriptionId, $invocationContext)
+		{
+			$this.SubscriptionId = $SubId
+			$this.OrganizationToScan = $OrgName
+			$this.PATToken = $PATToken
+			$this.ProjectNames = $Proj
+			$this.ExtendedCommand = $ExtCmd
+			$this.SetupComplete = $false
+			$this.ControlSettings = [ConfigurationManager]::LoadServerConfigFile("ControlSettings.json");
+			$this.ImageName = $this.ControlSettings.DockerImage.ImageName
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:107]: add check member

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+
+			#Step 2 : Check if Owner or Contributor role is available at subscription scope.
+			if ($null -eq $Context)  {
+				$output = "No Azure login found. Azure login context is required to setup Continuous Assurance."
+			}
+			else
+			{
+				if($Context.Subscription.SubscriptionId -ne $this.SubscriptionId)
+				{
+					$Context = set-azcontext -Subscription $this.SubscriptionId -Force | out-null
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:141]: what if subid is malformed?
it still throws an exception

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+				}
+				$Scope = "/subscriptions/"+$this.SubscriptionId
+				$RoleAssignment = @((Get-AzRoleAssignment -Scope $Scope -SignInName $Context.Account.Id -IncludeClassicAdministrators ).RoleDefinitionName | where {$_ -eq "Owner" -or $_ -eq "CoAdministrator" -or $_ -match "ServiceAdministrator"} )
+				if ($RoleAssignment.Count -eq 0)
+				{
+					Write-Host "Please make sure you have Owner or Contributor role on target subscription. If your permissions were elevated recently, please run the 'Disconnect-AzAccount' command to clear the Azure cache and try again."
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:147]: remove contributor

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+				{
+					Write-Host "Function app '$($this.FuncAppName)' created" -ForegroundColor Green
+					$this.CreatedResources += $FuncApp.Id
+				}
+				
+				#Step 5: Validate if AI got created
+				$AppInsight = Get-AzResource -Name $this.AppInsightsName -ResourceType Microsoft.Insights/components
+				if($null -eq $AppInsight) 
+				{
+					Write-Host "Application Insight '$($this.AppInsightsName)' creation failed" -ForegroundColor Red
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:237]: application insights

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+					$this.CreatedResources += $KeyVault.resourceid
+				}
+		
+		
+				#Step 8: Add PAT token secret to KeyVault
+				#$Secret = ConvertTo-SecureString -String $this.PATToken -AsPlainText -Force
+				$CreatedSecret = Set-AzKeyVaultSecret -VaultName $this.KeyVaultName -Name $this.SecretName -SecretValue $this.PATToken
+				if($null -eq $CreatedSecret) 
+				{
+					Write-Host "Secret creation in KeyVault failed" -ForegroundColor Red
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:281]: we can specify that secret for PAT is created

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+		
+		
+				#Step 9: Get Identity details of function app to provide access on keyvault and storage
+				$FuncApp = Get-AzWebApp -Name $this.FuncAppName -ResourceGroupName $this.RGname		
+				$FuncAppIdentity= $FuncApp.Identity.PrincipalId 						
+				$MSIAccessToKV = Set-AzKeyVaultAccessPolicy -VaultName $this.KeyVaultName -ResourceGroupName $this.RGname -PermissionsToSecrets get,list -PassThru -ObjectId $FuncAppIdentity
+				$IsMSIAccess = $MSIAccessToKV.AccessPolicies | ForEach-Object { if ($_.ObjectId -match $FuncAppIdentity ) {return $true }}
+				if($IsMSIAccess -eq $true) 
+				{
+					Write-Host "MSI access to keyvault provided" -ForegroundColor Green
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:296]: KeyVault

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+					Write-Host "App settings updated" -ForegroundColor Green
+				}
+		
+		
+				Write-Host "`r`nSetup Complete!" -ForegroundColor Green
+				Restart-AzFunctionApp -name $this.FuncAppName -ResourceGroupName $this.RGname -SubscriptionId $this.SubscriptionId -Force
+		
+				Write-Host "Scan will begin at $($this.ScanTriggerLocalTime)" -ForegroundColor Green
+				$this.SetupComplete = $true
+				$messageData += [MessageData]::new("The following resources were created in resource group $($this.RGName) as part AzureDevOps of Continuous Assurance", ($this.CreatedResources| Out-String))
```
Comment (abhaydaga) [inline src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:359]: part of AzSK.ADO CA

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 (hunk 2 above)
Comment (Garima-msft) [reply src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:62]: this.timestamp is of string type, cannot be reused here

File: src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1 (hunk 4 above)
Comment (Garima-msft) [reply src/AzSK.AzureDevOps/Framework/Core/ContinuousAssurance/CAAutomation.ps1:155]: Yes , below message with be on pwsh 
Error validating permissions on the subscription

**Pull request merged**